File: Задание1/prog.hs
(?) cond a = if cond then (\x -> a) else (\x -> x)

toLowEven :: Int -> Int
toLowEven x = (even x) ? x $ x-1

findDoubleString :: String -> Int -> Int -> Int
findDoubleString str i middle =
        if i > middle
        then
            if str !! (i-1) == str !! (i - middle - 1)
            then
                findDoubleString str (i-1) middle
            else
                findDoubleString str newI newMiddle
        else
            middle

    where newI = toLowEven (i-1)
          newMiddle = newI `div` 2

numberCopyOperation :: String -> Int
numberCopyOperation str = findDoubleString str i middle

    where i = toLowEven $ length str
          middle = i `div` 2

result :: String -> Int -> String
result "" _ = ""
result str num = result (take lenNextStr str) num ++ sep ++ str

    where len = length str
          sep = (len > 0) ? "->" $ " = "
          lenNextStr = (num * 2 == len) ? num $ len-1
          

main :: IO ()
main = do
    inputS <- getLine
    let s = inputS
    let num = numberCopyOperation s
    let operCnt = length s - num + ((num > 0) ? 1 $ 0)
    putStrLn $ "operations count: " ++ (show operCnt) ++ result s num

File: Задание1/prog.java
import java.util.Scanner;

public class Main {
    // возвращаемое значение функции,
    // представляет собой номер (нумерация с 0) операции копирования строки,
    // а так же кол-во копируемых символов

    public static int numberCopyOperation(String s) {
        int i = s.length() - (s.length() % 2);
        int middle = i / 2;

        while (i > middle) {
            if (s.charAt(i - 1) != s.charAt(i - middle - 1)) {
                i = (i % 2 != 0 ? i : i - 1);
                middle = i / 2;
            }
            i--;
        }

        return middle;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String s = scanner.nextLine();

        int num = numberCopyOperation(s);
        int operationsCnt = s.length() - num + (num != 0 ? 1 : 0);

        System.out.print("operations count: " + operationsCnt);

        for (int i = 0; i < operationsCnt; i++) {
            int lenOutStr = i + (i < num || num == 0 ? 1 : num);
            if (i == 0) {
                System.out.print(" = " + s.substring(0, lenOutStr));
            } else {
                System.out.print("->" + s.substring(0, lenOutStr));
            }
        }
    }
}